Expose a http.Client instance

chelnak opened the pull request:
Closes #13 

The idea behind this PR is that a developer can opt to configure another library (e.g go-github) with a http client instance that has been configured to use gh-cli auth.

chelnak committed:
Allow developers to return a http.Client instance

This commit adds a new top level function that returns a http.Client
instance that is pre-configured with the auth token from gh-cli.

The aim is to allow developers to use the configured http.Client
instance in third party packages such as go-github.

chelnak committed:
Exopose new method that creates a HTTPClient

Adds a new function that calls newHTTPClient with the given options.
This felt like right approach because it means we don't have to
change existing code or duplicate anything.

chelnak committed:
Ensure that request headers aren't overwritten

Previous tests showed that go-gh would overwrite any headers that
were already set on the request object. So this meant that when using
a third party library, their headers would be replaced potentially
leading to unintended behaviour.

mislav commented on example/httpclient/main.go:
```diff
@@ -0,0 +1,32 @@
+	"log"
+
+	"github.com/cli/go-gh"
+	"github.com/google/go-github/v42/github"
```
Let's not add this example because it adds a whole new library to our `go.mod` and it's entirely unnecessary. Go developers typically don't need an example of what to do with a `http.Client` instance.

chelnak replied:
👍

mislav commented on gh.go:
```diff
@@ -107,6 +108,35 @@ func GQLClient(opts *api.ClientOptions) (api.GQLClient, error) {
+		if err != nil {
+			return nil, err
+		}
+		opts.AuthToken = token
```
This `http.Client` instance will be configured with a `github.com` authentication token. However, a `http.Client` instance can be used to make arbitrary requests to any host, and we don't know how it will be used. We absolutely would *not* want that the `github.com` auth token is sent to non-`github.com` hosts. https://github.com/cli/go-gh/blob/6722e5635158d5b607f7d955253e826f7c87b5a4/internal/api/http.go#L219

If we were to expose a generic `http.Client`, we'd have to take precaution so that the token isn't set to other hosts. It might be as simple as only sending the authentication token with requests to hostnames that match `opts.Host`. What do you think?

chelnak replied:
Yeah this sounds sensible to me. I'll make a start by adding something that stops the auth token being forward to non `opts.Host` requests as you suggest.

Thanks!

chelnak replied:
@mislav I've made a small change that stops the authorization header from being included if the hosts don't match.

The interesting thing was that the configured host and request host will never be an explicit match (github.com vs api.github.com) so I am checking for existence of the one in the other.

Unsure if this is a viable solution so your feedback is very welcome!

chelnak replied:
I've added some tests for the above change.  👍

chelnak committed:
Validate request host

Validate that the configured host matches the requests host
so that we don't send the auth token to arbitrary locations.

If the request host does not contain the host configured in
opts. Host then we remove the authorization property from
the header map.

chelnak committed:
Add a new test case for host matching

The tests added in this commit aim to prove that if the requested
host does not match opts.Host the auth token is removed from
the request headers.

chelnak commented:
~I'm also wondering if I should move `NewHttpClient` from `http_client.go` in to `http.go`... it's not an interface with methods or anything so feel like it's a bit overkill having it in a new file.~

Resolved by exporting the internal method.

samcoe commented on internal/api/http_client.go:
```diff
@@ -0,0 +1,12 @@
+)
+
+func NewHTTPClient(opts *api.ClientOptions) *http.Client {
+	c := newHTTPClient(opts)
```
Seems like we could just export `newHTTPClient` rather than add this function.

chelnak replied:
No problemo, will get this done shortly.

samcoe commented on gh.go:
```diff
@@ -107,6 +108,35 @@ func GQLClient(opts *api.ClientOptions) (api.GQLClient, error) {
 	return iapi.NewGQLClient(opts.Host, opts), nil
 }
 
+// HTTPClient builds a client that can be passed to another library.
+// As part of the configuration a hostname, auth token, and default set of headers are resolved
+// from the gh environment configuration. These behaviors can be overridden using the opts argument.
```
We should add some comments about how providing the host option here will not actually direct a request to that host which is different from the other clients. Also, we should add some comments about how the auth token is protected from being sent to arbitrary hosts.

chelnak replied:
Sounds sensible to me. Commit incoming.

samcoe commented on internal/api/http.go:
```diff
@@ -232,12 +237,20 @@ func newHeaderRoundTripper(authToken string, headers map[string]string, rt http.
 func (hrt headerRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If request is not for the host we have configured, do not add authorization headers.
+	if !isValidHost(req.URL.Host, hrt.host) && hrt.headers[authorization] != "" {
+		delete(hrt.headers, "Authorization")
```
```suggestion
		delete(hrt.headers, authorization)
```

chelnak replied:
The key in this map is actually `Authorization`. Applying your suggestion would make the test fail.

map:

```go
map[Accept:application/vnd.github.merge-info-preview+json, application/vnd.github.nebula-preview, application/vnd.github.antiope-preview, application/vnd.github.shadow-cat-preview Authorization:token oauth_token Content-Type:application/json; charset=utf-8 Time-Zone:Europe/London User-Agent:go-gh]
```

failing test:

```go
--- FAIL: TestNewHTTPClientWithDifferentHost (0.00s)
    --- FAIL: TestNewHTTPClientWithDifferentHost/removes_authorization_header_for_a_different_host (0.00s)
        /Users/craigg/code/go-gh/internal/api/http_test.go:145: 
            	Error Trace:	http_test.go:145
            	Error:      	Not equal: 
            	            	expected: ""
            	            	actual  : "token oauth_token"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-
            	            	+token oauth_token
            	Test:       	TestNewHTTPClientWithDifferentHost/removes_authorization_header_for_a_different_host
FAIL
FAIL	github.com/cli/go-gh/internal/api	0.141s
FAIL
```

chelnak replied:
Ignore that. Completely overlooked that it was a const 😸

chelnak committed:
Expose api/NewHTTPClient

Previously an additional abstraction that exposed newHTTPClient
was added, this commit removes that and directly exports newHTTPClient.

Also all references have been updated along with some small modifications
to HTTPClient.

chelnak committed:
Add some further comments for HTTPClient

Resolving some PR comments and adding details about how opts.
Host will behave and how tokens are protected from being sent to
arbitrary hosts.

mislav commented on internal/api/http.go:
```diff
@@ -232,12 +237,20 @@ func newHeaderRoundTripper(authToken string, headers map[string]string, rt http.
 func (hrt headerRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If request is not for the host we have configured, do not add authorization headers.
+	if !isValidHost(req.URL.Host, hrt.host) && hrt.headers[authorization] != "" {
+		delete(hrt.headers, authorization)
```
Instead of deleting the `authorization` header for invalid hosts, let's conditionally add it only for valid hosts.

chelnak replied:
This is done (i think). I've also reworked the tests to align.

mislav commented on internal/api/http.go:
```diff
@@ -232,12 +237,20 @@ func newHeaderRoundTripper(authToken string, headers map[string]string, rt http.
 
 func (hrt headerRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If request is not for the host we have configured, do not add authorization headers.
+	if !isValidHost(req.URL.Host, hrt.host) && hrt.headers[authorization] != "" {
```
Please use `u.Hostname()` instead of `u.Host`, since the former strips out the port name if present

chelnak replied:
Thanks for this. I have modified as suggested.

mislav commented on internal/api/http.go:
```diff
@@ -187,16 +187,21 @@ func inspectableMIMEType(t string) bool {
 }
 
+func isValidHost(reqHost, hrtHost string) bool {
+	return strings.Contains(reqHost, hrtHost)
```
Hostnames in URLs are case-insensitive, so you will want to factor that in when performing comparisons.

To compare whether two values are identical hostnames, you can use `strings.EqualFold(host1, host2)`.

To check whether one hostname is a subdomain of the other, enforce the use of the `.` separator and do strict suffix check. For this, `strings.Contains()` is not a good method because it's imprecise: a host called `thegithub.company` does contain the string `github.com` but is *not* a subdomain of `github.com`. Instead, this is a better check: `strings.HasSuffix("api.github.com", ".github.com")`. Again, remember to normalize the case of input arguments before submitting them for the HasSuffix check.

chelnak replied:
I've made this more robust based on your suggestions.

mislav reviewed (changes requested):
Getting there!

chelnak committed:
Refactor request host validation

To address PR comments this commit changes how the authorization
header is added to the request. It will now only be added for
valid hosts. Additionally HasSuffix is used rather than Contains
to check whether x is a subdomain of y.

mislav commented on internal/api/http.go:
```diff
@@ -187,8 +187,17 @@ func inspectableMIMEType(t string) bool {
+		strings.HasSuffix(reqHost, "."+hrtHost) ||
+		// // Check if the reqHost is a subdomain of hrtHost
+		// // e.g strings.HasSuffix("github.com", "".api.github.com") == false
+		strings.HasSuffix(hrtHost, "."+reqHost)
```
Let's remove this check; I can't envision a scenario where it would be necessary

chelnak replied:
Removed

mislav commented on internal/api/http.go:
```diff
@@ -187,8 +187,17 @@ func inspectableMIMEType(t string) bool {
 
-func isValidHost(reqHost, hrtHost string) bool {
-	return strings.Contains(reqHost, hrtHost)
+func isRequestHostValid(reqHost, hrtHost string) bool {
```
Since this function checks whether the request host is under a specific domain (e.g. `github.com`), and not whether a hostname is "valid" (which is vague and unspecified), how about renaming it to `isSameDomain()`?

chelnak replied:
Good idea - I've renamed as suggested

mislav commented on internal/api/http.go:
```diff
@@ -187,8 +187,17 @@ func inspectableMIMEType(t string) bool {
+	hrtHost = strings.ToLower(hrtHost)
+
+	return (reqHost == hrtHost) ||
+		// // Check if the hrtHost is a subdomain of reqHost
+		// // e.g strings.HasSuffix("api.github.com", "".github.com") == true
```
This comment is wrong and full of typos. You can remove it; it should be easy to understand what the logic is doing here.

chelnak replied:
Whoops - the comments should have been removed. They were for my reference when debugging.

mislav commented on internal/api/http.go:
```diff
@@ -187,8 +187,17 @@ func inspectableMIMEType(t string) bool {
 
-func isValidHost(reqHost, hrtHost string) bool {
-	return strings.Contains(reqHost, hrtHost)
+func isRequestHostValid(reqHost, hrtHost string) bool {
```
Let's avoid acronyms and abbreviations in variable names when not necessary.
- `reqHost` could be `requestHost`
- `hrtHost` (what is "hrt" in this context anyway?) could simply be `domain`

chelnak replied:
Yep can see where you are coming from here and have made the suggested changes.

mislav reviewed (dismissed):
Thanks for the updates!

chelnak committed:
Refactor domain checking function

To address PR comments:
* isRequestHostValid has been renamed to isSameDomain
* reqHost has been renamed to request host
* hrtHost has been renamed to domain
* Reverse HasPrefix check has been removed
* Debug comments have been removed

samcoe committed:
Extract out option resolving code into its own function

samcoe committed:
Simplify logic around when to add authorization header

samcoe committed:
Consolidate tests a bit

samcoe reviewed (approved):
@chelnak This looks great thanks for making the requested changed. I pushed just a couple commits refactoring out a common `resolveOptions` function and adding a bit of 💅

chelnak commented:
@samcoe Thank you - appreciate the guidance from you both. 😄

chelnak commented:
Hey! Is there anything more I can do for this PR?

chelnak commented:
Thank you ❤️ !
